Refactoring and modularising of frontend done

photomz opened the pull request:
Please reflect on changes made on the markus branch and verify for merging with master

101arrowz commented:
I made a new commit with some changes that I wanted to make. Please take a look at it @photomz.

101arrowz commented on src/components/ConjugationContainer/data.json:
```diff
@@ -0,0 +1,10 @@
+		"chess queen",
+		"chess pawn"
+	]
+}
```
Consider renaming this file

101arrowz commented on src/components/SearchBar.jsx:
```diff
@@ -0,0 +1,91 @@
+  },
+);
+
+SearchBar.displayName = 'SearchBar';
```
Is this necessary?

photomz replied:
Eslint yells at me otherwise

101arrowz commented on src/global/reset.css:
```diff
@@ -0,0 +1,446 @@
+/* make sure to set some focus styles for accessibility */
+:focus {
+  outline: 0;
+}
```
Consider removing this because it's bad for accessibility (it even mentions to set focus styles for accessibility, but we don't do that).

photomz replied:
Could you elaborate on why it will be bad for accessibility?

101arrowz replied:
If the browser has a default to outline an element when it's focused (i.e. press tab to highlight it) and you're removing it, the user cannot see the outline and can't tell what's highlighted. This really sucks for people who have trouble using a mouse and use their keyboard instead.

101arrowz commented on src/global/routes.js:
```diff
@@ -0,0 +1,11 @@
+  About: '/about',
+  Conjugate: '/conjugate',
+  GitHub: '/github',
+};
```
Consider being more consistent between JS and JSON files for static data. I'd suggest using only `.js` files.

photomz replied:
Noted, will have all static data be <code>.js</code> files.

101arrowz commented on src/hooks/useOutsideClick.js:
```diff
@@ -0,0 +1,34 @@
+  }, []);
+};
+
+export default useOutsideClick;
```
This seems excessive for our purposes. We should consider replacing this with a better solution later (adding too many event listeners to `document` will make the page unnecessarily inefficient).

photomz replied:
We will only use this hook for the navbar - not for all components on all pages - so it will not have a major impact on performance

101arrowz commented on src/pages/Conjugate/index.jsx:
```diff
@@ -0,0 +1,123 @@
+  const handleFilterResults = value => filterVerbs(value, 5);
+
+  const handleSearchClick = value => {
+    value = decodeURI(value).toLowerCase();
```
Why are we decoding the value provided by the searchbar?

photomz replied:
Because it was there in the original code and I thought it might break something...you can remove it if wanted

101arrowz commented on src/pages/GitHub.jsx:
```diff
@@ -0,0 +1,26 @@
+import styled from 'styled-components';
+
+const StyledSegment = styled(Segment)`
+  margin: 25vh 25vw;
```
This styling doesn't work for me, at least on my machine. I'm fairly certain that it will work better if we use:
```css
position: absolute;
top: 50%;
left: 50%;
transform: translate(-50%, -50%);
```
or something similar.

photomz replied:
Noted, will make changes.

photomz commented on src/pages/Conjugate/index.jsx:
```diff
@@ -28,16 +28,19 @@ const Conjugate = () => {
-    // Prevent memory leak by aborting fetch when component unmounts
-    const abortController = new AbortController();
-    const { signal } = abortController;
-
```
The abort controller needs to be kept to prevent a memory leak when the component unmounts - please do not remove

photomz commented on src/pages/Conjugate/index.jsx:
```diff
@@ -28,16 +28,19 @@ const Conjugate = () => {
     if (value === verb) return;
     setSearchValue(value);
-    setAction('redirect');
+    history.push(`/conjugate/${value}`);
```
Why are we still using the history object?

photomz commented:
@101arrowz I made a new commit - I did not touch any changes except I changed history back to redirect and added back AbortController.

The reasoning for <code><Redirect/></code> instead of <code>history.push()</code> goes that react-router under the hood does **not** rerender components on a redirect if the component rendered stays the same.

<code>AbortController</code> is added to prevent a memory leak during <code/>componentWillUnmount</code>, and a setTimeout does not prevent the memory leak, unfortunately.

Other than that, feel free to merge.

101arrowz commented:
I've reverted part of your previous commit.

Under the hood, the `Redirect` component calls `history.push` anyway. You can see this in the [source code](https://github.com/ReactTraining/react-router/blob/master/packages/react-router/modules/Redirect.js#L21). Did you notice that I didn't import the `history` package, that `history` was passed in props? `react-router` provides the React components it renders a `history` prop for a reason; they __recommend__ using `history` for programmatic routing without a rendered component. Your solution with `Redirect` does the exact same thing, but the code looks ugly and it is inefficient, since it requires a rerender by React to remove the current view and replace it with `null` (it won't be the same route when you search a new verb, so yes, it will rerender). More importantly, it causes a blank screen during the transition state. If you look for an invalid verb, the screen is blank instead of showing you the "Invalid verb" error!

`AbortController` is not polyfilled by Babel and has a relatively low amount of support. I created that solution with a flag and `setTimeout` for that reason. As far as I know, React suggests using either an abortable `fetch` or a flag. However, I realize that completing the `fetch` call but not settting state because of a flag still results in a memory leak, even if React doesn't report it. That's why I haven't changed it back. I wish it were more widely supported, though!
